src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,594 @@
+            parsePos.setIndex(source.length());
+        }
+
+        if (parsed instanceof Object[] objs) {
+            parsePos.setErrorIndex(-1);
+            return Arrays.asList(objs);
+        } else {
+            // MessageFormat.parseObject() failed
+            return null;
+        }
+    }
+
+    @Override
+    public AttributedCharacterIterator formatToCharacterIterator(Object arguments) {
+        Objects.requireNonNull(arguments);
+
+        if (arguments instanceof List<?> objs) {
+            var a = objs.toArray(new Object[0]);
+            return generateMessageFormat(a).formatToCharacterIterator(a);
+        } else if (arguments instanceof Object[] objs) {
+            return generateMessageFormat(objs).formatToCharacterIterator(objs);
+        } else {
+            throw new IllegalArgumentException("The arguments should be a List<Object> or an Object[]");
+        }
+    }
+
+    /**
+     * Checks if this {@code ListFormat} is equal to another {@code ListFormat}.
+     * The comparison is based on the {@code Locale} and formatting patterns, given or
+     * generated with {@code Locale}, {@code Type}, and {@code Style}.
+     * @param obj the object to check, {@code null} returns {@code false}
+     * @return {@code true} if this is equals to the other {@code ListFormat}
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj instanceof ListFormat other) {
+            return locale.equals(other.locale) &&
+                Arrays.equals(patterns, other.patterns);
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hash(locale, Arrays.hashCode(patterns));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return
+            """
+            ListFormat [locale: "%s", start: "%s", middle: "%s",  end: "%s", two: "%s", three: "%s"]
+            """.formatted(locale.getDisplayName(), patterns[START], patterns[MIDDLE], patterns[END], patterns[TWO], patterns[THREE]);
+    }
+
+    private MessageFormat generateMessageFormat(Object[] input) {
+        var len = input.length;
+        return switch (len) {
+            case 0 -> throw new IllegalArgumentException("There should at least be one input string");
+            case 1 -> new MessageFormat("{0}", locale);
+            case 2, 3 -> new MessageFormat(patterns[len + 1], locale);
+            default -> new MessageFormat(createMessageFormatString(len), locale);
+        };
+    }
+
+    private String createMessageFormatString(int count) {
+        var sb = new StringBuilder(256).append(patterns[START]);
+        IntStream.range(2, count - 1).forEach(i -> sb.append(middleBetween).append("{").append(i).append("}"));
+        sb.append(patterns[END].replaceFirst("\\{0}", "").replaceFirst("\\{1}", "\\{" + (count - 1) + "\\}"));
```

Comment:
From what it looks, it could be a concern for potentially adding large number of long strings with a list of small items. I don't seem to see where the input is limited.

Reply:
Good point. Will add some kind of limitation.

Reply:
I can't see what arbitrary limit would make sense.
But perhaps an APINote that  formatting the string from an excessively long list may exceed memory or string sizes. (Without being specific about the exception being thrown).

Reply:
OK, will ad that @apiNote in the 1-arg getInstance()